### ISSUE TITLE
Unable to install devel if stable installed.

### ISSUE THREAD

**Issue by vladshablinsky** [Homebrew/brew#1162]
## Description

```
➜ vlad:Homebrew$ brew install ack                                        master
==> Using the sandbox
==> Downloading http://beyondgrep.com/ack-2.14-single-file
Already downloaded: /Users/vlad/Library/Caches/Homebrew/ack-2.14.14-single-file
==> pod2man /usr/local/Cellar/ack/2.14/bin/ack ack.1
🍺  /usr/local/Cellar/ack/2.14: 3 files, 182.9K, built in 1 second
➜ vlad:Homebrew$ brew install ack --devel                                master
Warning: ack-2.14 already installed
```
## Workaround

We'd probably want to warn user that `stable` keg is linked and suggest user unlinking it. After the keg have been unlinked we should be able to install any version that is newer than the latest version installed in the `Cellar` e.g. `HEAD` or `devel`.
## TODO
- [ ]  WIP Improve `install` #1343
- [ ]  Improve `upgrade`

CC @Homebrew/maintainers

**Comment by MikeMcQuaid** [comment]
Seems sensible but out of interest what happens if you use `brew upgrade ack --devel`?

**Comment by vladshablinsky** [comment]
> Seems sensible but out of interest what happens if you use brew upgrade ack --devel?

Wow, it works! Will experiment with it a bit and then close the issue if I don't find anything strange. Thanks!

**Comment by vladshablinsky** [comment]
Seems that the warning message should be updated, though.

**Comment by ilovezfs** [comment]
Yup, that works, though it does still block any transition from HEAD to devel or devel to stable or HEAD to stable.

**Comment by MikeMcQuaid** [comment]
I think requiring intervention on a downgrade is sensible but agreed the messaging should be updated in both cases.

**Comment by ilovezfs** [comment]
What intervention actually works for a downgrade currently other than uninstall?

**Comment by MikeMcQuaid** [comment]
`unlink`, I think?

**Comment by ilovezfs** [comment]
Nope!

**Comment by ilovezfs** [comment]
```
iMac-TMP:~ joe$ brew install --devel ack
Warning: ack-HEAD-f0ad2b9 already installed
iMac-TMP:~ joe$ brew unlink ack
Unlinking /usr/local/Cellar/ack/HEAD-f0ad2b9... 2 symlinks removed
iMac-TMP:~ joe$ brew install --devel ack
Warning: ack-HEAD-f0ad2b9 already installed, it's just not linked
```

**Comment by MikeMcQuaid** [comment]
Interesting. No idea, then!

**Comment by ilovezfs** [comment]
Yeah... I guess we could have a `downgrade` command :scream_cat:

**Comment by ilovezfs** [comment]
Current behavior for an upgrade (stable -> devel or head, or devel -> head)
unlink + install -> fail
upgrade -> succeed, and prior version is kept
uninstall + install -> succeed, but prior version is removed due to the uninstall

Current behavior for a downgrade (head -> devel or stable, or devel -> stable)
unlink + install -> fail
downgrade -> command doesn't exist
uninstall + install -> succeed but prior version is removed due to the uninstall

I think we probably want to change this to ...

New behavior for an upgrade (stable -> devel or head, or devel -> head)
unlink + install -> succeed, and prior version is kept
upgrade -> succeed, and prior version is kept
uninstall + install -> succeed, but prior version is removed due to the uninstall

New behavior for a downgrade (head -> devel or stable, or devel -> stable)
unlink + install -> succeed, and prior version is kept
downgrade -> command doesn't exist (consider creating new command, but probably reject the idea)
uninstall + install -> succeed but prior version is removed due to the uninstall

**Comment by vladshablinsky** [comment]
> Yeah... I guess we could have a downgrade command 🙀

Downgrade seems quite strange for me as a command.
Probably it'd be nice if we could upgrade specific spec passing the corresponding flag for it regardless the latest spec installed.

**Comment by MikeMcQuaid** [comment]
> unlink + install -> succeed, and prior version is kept
> upgrade -> succeed, and prior version is kept
> uninstall + install -> succeed, but prior version is removed due to the uninstall

Makes sense to me 👍

**Comment by ilovezfs** [comment]
Note that there may be a reason not to implement this. It will lead to a reduction in safety for keg_only formulae, since no `unlink` step will be required, and devel will silently displace an installed stable version simply by running `brew install --devel foo`, potentially breaking dependents, so this may require more thought unless that's an acceptable safety regression.

**Comment by vladshablinsky** [comment]
> devel will silently displace an installed stable version simply by running brew install --devel foo

@ilovezfs the only difference with non-keg-only formula is that it happens silently? I mean `opt` is overwritten anyway after `instal`l regardless whether the formula is keg-only or not.
Am I missing something?

**Comment by ilovezfs** [comment]
@vladshablinsky the concern is only that in the keg-only case the users may not realize they're about to replace an existing install, since there's no manual step requiring them to unlink anything first.

**Comment by vladshablinsky** [comment]
@ilovezfs in this case we can try to use some interactivity for `keg_only` e.g. we can ask if user really wants to install new version and expect user to enter `y` or `n`.

**Comment by MikeMcQuaid** [comment]
@vladshablinsky Need to figure something out without interactivity as it's something we don't do in Homebrew.

**Comment by vladshablinsky** [comment]
@MikeMcQuaid @ilovezfs well, we can warn and ask user to use `--force` or something for `keg_only` formula.

**Comment by vladshablinsky** [comment]
> well, we can warn and ask user to use --force or something for keg_only
Actually we can just show a message like `use brew upgrade <formula>` to install new version.

**Comment by MikeMcQuaid** [comment]
@vladshablinsky Sounds good.

**Comment by ilovezfs** [comment]
That won't allow the downgrades for keg_only. Force is probably a more flexible answer...

**Comment by vladshablinsky** [comment]
@ilovezfs do we allow downgrades at all?

**Comment by ilovezfs** [comment]
You'd want to be able to install stable alongside HEAD without uninstalling HEAD first, just like you'd want to be able to install HEAD alongside stable without uninstalling stable.

Technically none of these are upgrades or downgrades, just installing multiple specs.

**Comment by vladshablinsky** [comment]
I've already addressed the idea in this comment https://github.com/Homebrew/brew/issues/1162#issuecomment-249807300

Here it is: 
> Probably it'd be nice if we could upgrade specific spec passing the corresponding flag for it regardless the latest spec installed.

So we pass the spec we want explicitly and Homebrew installs the latest version of that spec.

**Comment by ilovezfs** [comment]
Personally I like that idea because overloading upgrade for a spec switch stable -> devel or devel -> HEAD seems like an abuse of what upgrade means, and the fact that it currently works I think is actually a bug not a feature.

**Comment by vladshablinsky** [comment]
> Personally I like that idea 

You mean the idea adding `downgrade` command or the idea with explicit specs?

Upgrade without argument should probably behave same way it does now. If we introduced `--stable` option to define spec explicitly, we'd be able to use `brew install [--stable|--devel|--HEAD]` and
`brew upgrade [--stable|--devel|--HEAD]`.

**Comment by ilovezfs** [comment]
I mean the explicit specs. Yes, I think you have the right idea.

**Comment by vladshablinsky** [comment]
@ilovezfs 
> Note that there may be a reason not to implement this. It will lead to a reduction in safety for keg_only formulae, since no unlink step will be required, and devel will silently displace an installed stable version simply by running brew install --devel foo, potentially breaking dependents, so this may require more thought unless that's an acceptable safety regression.

> the concern is only that in the keg-only case the users may not realize they're about to replace an existing install, since there's no manual step requiring them to unlink anything first.

@ilovezfs correct me if I'm wrong, but the issue you addressed is already present. What if `keg_only` gets outdated? 

```ruby
# cmd/install.rb#L138-L156

        current = f if f.installed?
        current ||= f.old_installed_formulae.first

        if current
          msg = "#{current.full_name}-#{current.installed_version} already installed"
          unless current.linked_keg.symlink? || current.keg_only?
            msg << ", it's just not linked"
          end
          opoo msg
        elsif f.migration_needed? && !ARGV.force?
          # Check if the formula we try to install is the same as installed
          # but not migrated one. If --force passed then install anyway.
          opoo "#{f.oldname} already installed, it's just not migrated"
          puts "You can migrate formula with `brew migrate #{f}`"
          puts "Or you can force install it with `brew install #{f} --force`"
        else
          formulae << f
        end
```
https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/install.rb#L138-L156

Should be fixed in #1343

**Comment by MikeMcQuaid** [comment]
Fixed!
